core/common/src/main/java/alluxio/Configuration.java

```diff
@@ -406,10 +407,34 @@ public static Duration getDuration(PropertyKey key) {
   }
 
   /**
-   * @return a view of the internal {@link Properties} of as an immutable map
+   * @return a map of the resolved properties(include default properties)
    */
   public static Map<String, String> toMap() {
-    return Collections.unmodifiableMap(PROPERTIES);
+    Map<String, String> map = toRawMap();
+    // Resolve values with ${VALUE} format
+    String nestedPropIdentifier = "$";
+    for (Map.Entry<String, String> entry : map.entrySet()) {
+      String value = entry.getValue();
+      if (value.contains(nestedPropIdentifier)) {
+        map.put(entry.getKey(), lookup(value));
+      }
+    }
+    return map;
+  }
+
+  /**
+   * @return a map of the raw properties(include default properties)
+   */
+  public static Map<String, String> toRawMap() {
+    Map<String, String> map = new HashMap<>(PROPERTIES);
+    for (PropertyKey key : PropertyKey.defaultKeys()) {
+      String propName = key.getName();
+      String defaultValue = key.getDefaultValue();
+      if (!map.containsKey(propName) && defaultValue != null) {
```

Comment:
why do we exclude properties that have null values?

Reply:
The first priority is usability by users who might inspect configuration through the UI or through the shell.

One use case is when a user thinks they have set a configuration property, but they didn't do it right. In that case, it's useful for them to be able to use the web or CLI interface to see

alluxio.master.hostname = null (source = default)

Then they know for sure that alluxio.master.hostname is not set. If we exclude it because it's null, they would need to know about that convention to understand that the property is not set.

On the other hand, if a user is looking for a non-null property, now they have more `null` properties to search through. However, I think we already have so many properties with non-null defaults that including nulls only adds 20-30% more properties. Users will already need to use grep or control-F to search for whatever property they are looking for.

Regarding the NPE issue, we have the same problem for all usage of `Configuration.get(key)` - calling it on a key that has a null value will also throw a runtime exception. Because we allow properties to be null, any code that works with properties needs to be aware that any property might be null. To help prevent NPEs we can include a mention in the javadoc that null is a valid value for keys in the map.

Reply:
Thanks for the explanation. Property keys with null value do have values for our customers.
When we decide to roll back to the previous way of merging default properties at the beginning, if we add null values, we will get NPE.

defaultProps.setProperty(key.toString(), value); //alluxio.Configuration.createDefaultProps(Configuration.java:143)
Caused by: java.lang.NullPointerException
	at java.util.Hashtable.put(Hashtable.java:460)
	at java.util.Properties.setProperty(Properties.java:166)
	at alluxio.Configuration.createDefaultProps(Configuration.java:143)

Reply:
HashTable do not allow null key or null value here.

Reply:
@LuQQiu @aaudiber  I see a lot of complication introduced due to non-supported null value in `HashMap` and `ConcurrentHashMap`. To really fix this, how about introducing a new static constant object `PropertyKey.NULL_VALUE` (whose `toString` can be "Not Set" or so), and assign this value to all properties which have no default value. In this way, I think we can process all property keys in a consistent way and reduce the code complexity.

Reply:
I think actually this is bug that we should fix even in 1.7. With 1.7.1, we can not list configuration with null default value, such as `alluxio.{logserver,worker}.hostname` which has no specified default value by `alluxio getConf`

Reply:
I agree to use something else representing having no default value than null. 

Could we use a single string NULL_VALUE, and setDefaultValue = NULL_VALUE?
```
private static final String NULL_VALUE = "Not Set"

  public static final PropertyKey ZOOKEEPER_ADDRESS =
      new Builder(Name.ZOOKEEPER_ADDRESS)
          .setDefaultValue(NULL_VALUE) 
          .setDescription("Address of ZooKeeper.")
          .setConsistencyCheckLevel(ConsistencyCheckLevel.ENFORCE)
          .build();
```

Reply:
I would not use a string `NULL_VALUE` of `"Not Set"`to avoid in case this value is a legit property value (who knows). We can create a static constant object `NO_DEFAULT_VALUE`

```java
  private static final Object NO_DEFAULT_VALUE = new Object() {
    @Override
    public String toString() {
      return "no default value set";
    }
  };
```

In this way, we can ensure it will not collide with legit value.

In addition, we can make this the default for `PropertyKey.mDefaultValue` to further minimize the code change

Reply:
I could have a fix in v1.7, and migrate it to master.

Reply:
Great idea! I agree to have an object representing no default value and fix that in another PR.